working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
 
-**TODO: Allow a `...` element in order to match suffixes or ignore extra
-elements. Allow capturing the rest in a variable.**
+    ```dart
+    // Variable:
+    (untyped: untyped, typed: int typed)
+    (:untyped, :int typed)
 
-#### Map binder
+    // Null-check and null assert:
+    (checked: checked?, asserted: asserted!)
+    (:checked?, :asserted!)
 
-A map binder access values by key from objects that implement `Map`.
+    // Cast:
+    (field: field as int)
+    (:field as int)
+    ```
 
-```
-mapBinder ::= mapTypeArguments? '{' mapBinderEntries '}'
+**TODO: Allow a `...` element in order to ignore some positional fields while
+capturing the suffix.**
 
-mapTypeArguments ::= '<' typeOrBinder ',' typeOrBinder '>'
+### Extractor pattern
 
-mapBinderEntries ::= mapBinderEntry ( ',' mapBinderEntry )* ','?
-mapBinderEntry ::= expression ':' binder
+```
+extractorPattern ::= extractorName typeArguments? '(' patternFields? ')'
+extractorName    ::= typeIdentifier | qualifiedName
 ```
 
-If it is a compile-time error if any of the entry key expressions are not
-constant expressions. It is a compile-time error if any of the entry key
-expressions evaluate to equivalent values.
+An extractor matches values of a given named type and then extracts values from
+it by calling getters on the value. Extractor patterns let users destructure
+data from arbitrary objects using the getters the object's class already
+exposes.
 
-#### Record binder
+This pattern is particularly useful for writing code in an algebraic datatype
+style. For example:
 
-A record pattern destructures fields from records and objects.
+```dart
+class Rect {
+  final double width, height;
 
-```
-recordBinder        ::= '(' recordFieldBinders ')'
+  Rect(this.width, this.height);
+}
 
-recordFieldBinders  ::= recordFieldBinder ( ',' recordFieldBinder )* ','?
-recordFieldBinder   ::= ( identifier ':' )? binder
-                      | identifier ':'
+display(Object obj) {
+  switch (obj) {
+    case Rect(width: w, height: h): print('Rect $w x $h');
+    default: print(obj);
+  }
+}
 ```
 
-Each field is either a binder which destructures a positional field, or a binder
-prefixed with an identifier and `:` which destructures a named field.
-
-Each named field in the record pattern is matched by calling a corresponding
-getter on the matched object. A record pattern can be applied to an object of
-any type with any set of named field patterns as long as the object being
-matched has getters with those names. (Positional fields, however, can only be
-matched by record values.)
-
-When destructuring named fields, it's common to want to bind the resulting value
-to a variable with the same name. As a convenience, the binder can be omitted on
-a named field. In that case, the field implicitly contains a variable binder
-subpattern with the same name. These are equivalent:
+You can also use an extractor to both match an enum value and destructure
+fields from it:
```

Comment:
Doesn't seem so from the grammar above, so is this really intended?

It seems to be mixing the constant pattern and the extractor pattern.

Can other constant patterns use it as well, or is it only `enum` declared constants?
(So not user-defined enum-like classes?)